src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java

```diff
@@ -732,14 +755,26 @@ protected SCMRevision retrieve(@NonNull final String revision, @NonNull final Ta
             if (name.startsWith(Constants.R_HEADS) && revision.equalsIgnoreCase(rev)) {
                 listener.getLogger().printf("Found match: %s revision %s%n", name, rev);
                 // WIN!
-                return new SCMRevisionImpl(new SCMHead(StringUtils.removeStart(name, Constants.R_HEADS)), rev);
+                return new GitBranchSCMRevision(new GitBranchSCMHead(StringUtils.removeStart(name, Constants.R_HEADS)), rev);
             }
             if (name.startsWith(Constants.R_TAGS) && revision.equalsIgnoreCase(rev)) {
                 listener.getLogger().printf("Candidate match: %s revision %s%n", name, rev);
                 // WIN but let's see if a branch also matches as that would save a fetch
                 fullTagMatches.add(name);
                 continue;
             }
+            if ("HEAD".equals(name)) {
```

Comment:
shouldn't we move this right up the top of the loop then? (or provide a comment as to why we can delay until here)